### PR TITLE
Added encryption for MXRealmCryptoStore

### PR DESCRIPTION
**Pull request by gileluard**

### CODE REVIEW

File: MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m [hunk 1]
```diff
@@ -1389,208 +1373,179 @@ + (RLMRealm*)realmForUser:(NSString*)userId andDevice:(NSString*)deviceId
...
+            NSLog(@"[MXRealmCryptoStore] deleting previous DB");
+            [RLMRealm deleteFilesForConfiguration:config error:&error];
+            if (error)
+            {
+                NSLog(@"[MXRealmCryptoStore] failed to delete previous DB: %@", error);
+            }
+        }
+    }
+    
+    NSLog(@"[MXRealmCryptoStore] realmForUser: end after %fs", [[NSDate date] timeIntervalSinceDate:startDate]);
```
Comment (ismailgulek) [inline MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m:1458]:
```suggestion
    NSLog(@"[MXRealmCryptoStore] realmForUser: end after %.3fms", [[NSDate date] timeIntervalSinceDate:startDate] * 1000);
```

Let's be consistent across timing values.

**Review by ismailgulek — CHANGES_REQUESTED** [review]

File: MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m [hunk 2]
```diff
@@ -1301,65 +1308,42 @@ + (RLMRealm*)realmForUser:(NSString*)userId andDevice:(NSString*)deviceId
...
-    // Check for a potential application group id.
-    NSString *applicationGroupIdentifier = [MXSDKOptions sharedInstance].applicationGroupIdentifier;
+
     if (applicationGroupIdentifier)
     {
         // Use the shared db file URL.
-        NSURL *sharedContainerURL = [[NSFileManager defaultManager] containerURLForSecurityApplicationGroupIdentifier:applicationGroupIdentifier];
-        NSURL *realmFileFolderURL = [sharedContainerURL URLByAppendingPathComponent:kMXRealmCryptoStoreFolder];
-        NSURL *realmFileURL = [[realmFileFolderURL URLByAppendingPathComponent:userId] URLByAppendingPathExtension:@"realm"];
+        NSURL *sharedFileURL = [self sharedFileURLWithApplicationGroupIdentifier:applicationGroupIdentifier FileName:realmFile];
```
Comment (SBiOSoftWhare) [inline MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m:1333]:
```suggestion
        NSURL *sharedFileURL = [self sharedFileURLWithApplicationGroupIdentifier:applicationGroupIdentifier fileName:realmFile];
```

File: MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m [hunk 3]
```diff
@@ -24,11 +24,16 @@
 #import "MXSession.h"
 #import "MXTools.h"
 #import "MXCryptoTools.h"
+#import "MXKeyProvider.h"
+#import "MXRawDataKey.h"
 
 NSUInteger const kMXRealmCryptoStoreVersion = 13;
 
-static NSString *const kMXRealmCryptoStoreFolder = @"MXRealmCryptoStore";
+NSString *const kMXRealmCryptoStoreDataType = @"org.matrix.sdk.MXRealmCryptoStoreDataType";
```
Comment (SBiOSoftWhare) [inline MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m:32]: For external constants it is recommended to not use the `k` prefix: `MXRealmCryptoStoreDataType`

**Review by SBiOSoftWhare — APPROVED** [review]

File: MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m [hunk 4]
```diff
@@ -1389,208 +1373,179 @@ + (RLMRealm*)realmForUser:(NSString*)userId andDevice:(NSString*)deviceId
...
+/**
+ Allow to know if the current DB is not encrypted and should be moved to an encrypted one.
+ 
+ @param fileName file name of the current DB file
+ @param applicationGroupIdentifier the Application Group Identifier if set. Nil otherwise.
+ @param defaultRealmPathURL default path of the Realm DB
+ 
+ @return YES if the DB should be moved to a encrypted DB. NO otherwise.
+ */
++ (BOOL)shouldMoveToEncryptedDBWithEncryptedFilename:(NSString *)fileName applicationGroupIdentifier:(NSString *)applicationGroupIdentifier defaultRealmPathURL:(NSURL *)defaultRealmPathURL
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m]: Thanks for the `_Nullable`. Looks like we still lack the check inside the method. `-[NSFileManager containerURLForSecurityApplicationGroupIdentifier:applicationGroupIdentifier]` will cause a crash for `applicationGroupIdentifier`.

File: MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m [hunk 5]
```diff
@@ -1301,65 +1308,42 @@ + (RLMRealm*)realmForUser:(NSString*)userId andDevice:(NSString*)deviceId
...
-            else
-            {
-                // Remove the residual db file.
-                [NSFileManager.defaultManager removeItemAtURL:defaultRealmFileURL error:nil];
-            }
-        }
-        else
+        NSError *error = nil;
+        
+        [self moveToShareContainerFrom:defaultRealmFileURL to:sharedFileURL error:&error];
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Crypto/Data/Store/MXRealmCryptoStore/MXRealmCryptoStore.m]: This seems better, thanks.

**Review by ismailgulek — APPROVED** [review]

**Review by manuroe — CHANGES_REQUESTED** [review]
We need to check if Realm can support multi process when encrypted.

**Comment by manuroe** [comment]
Closed in favour of https://github.com/matrix-org/matrix-ios-sdk/pull/1027

**Pull request closed without merging**